### PR TITLE
[query] Refactor function registry to not use overloading

### PR DESCRIPTION
**Pull request by tpoterba**
The overloading made development a huge headache. I think we should consider overloading more than ~2-way bad practice and try to avoid it.

### CODE REVIEW

**Comment by akotlar** [comment]
Yes, it is much harder to work with than necessary

File: hail/src/main/scala/is/hail/expr/ir/functions/Functions.scala [hunk 1]
```diff
@@ -377,8 +352,8 @@ abstract class RegistryFunctions {
     })
   }
 
-  def registerScalaFunction(mname: String, argTypes: Array[Type], rType: Type, pt: (Type, Seq[PType]) => PType)(cls: Class[_], method: String) {
-    registerCode(mname, argTypes, rType, pt) { case (r, rt, args) =>
+  def registerScalaFunction(mname: String, argTypes: Array[Type], rType: Type, pt: (Type, Seq[PType]) => PType, typeParams: Array[Type] = Array.empty)(cls: Class[_], method: String) {
```
Comment (akotlar) [inline hail/src/main/scala/is/hail/expr/ir/functions/Functions.scala:355]: not a big deal, but we don't need typeParams here, no registerScalaFunction passes this and registerCode now has a default.

File: hail/src/main/scala/is/hail/expr/ir/functions/Functions.scala [hunk 2]
```diff
@@ -395,147 +370,141 @@ abstract class RegistryFunctions {
...
-  def registerCode[A1](mname: String, mt1: Type, rt: Type, pt: (Type, PType) => PType)(impl: (EmitRegion, PType, (PType, Code[A1])) => Code[_]): Unit =
+  def registerCode1[A1](mname: String, mt1: Type, rt: Type, pt: (Type, PType) => PType)(impl: (EmitRegion, PType, (PType, Code[A1])) => Code[_]): Unit =
     registerCode(mname, Array(mt1), rt, unwrappedApply(pt)) {
-      case (r, rt, Array(a1: (PType, Code[A1]) @unchecked)) => impl(r, rt, a1)
+      case (r, rt, _, Array(a1: (PType, Code[A1]) @unchecked)) => impl(r, rt, a1)
     }
 
-  def registerCode[A1, A2](mname: String, mt1: Type, mt2: Type, rt: Type, pt: (Type, PType, PType) => PType)
+  def registerCode1t[A1](mname: String, typeArg: Type, mt1: Type, rt: Type, pt: (Type, PType) => PType)(impl: (EmitRegion, PType, Type, (PType, Code[A1])) => Code[_]): Unit =
+    registerCode(mname, Array(mt1), rt, unwrappedApply(pt)) {
```
Comment (akotlar) [inline hail/src/main/scala/is/hail/expr/ir/functions/Functions.scala:418]: registerCode should take Array(typeArg) in the last position.

Comment (akotlar) [reply hail/src/main/scala/is/hail/expr/ir/functions/Functions.scala:418]: can also be named typeArg1 for consistency

Comment (akotlar) [reply hail/src/main/scala/is/hail/expr/ir/functions/Functions.scala:418]: and we should probably name all of these either typeArg as before, or typeParam as you did above.

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/functions/Functions.scala:418]: good catch, thanks

File: hail/src/main/scala/is/hail/expr/ir/functions/Functions.scala (hunk 1 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/functions/Functions.scala]: good point.

File: hail/src/main/scala/is/hail/expr/ir/functions/Functions.scala (hunk 2 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/functions/Functions.scala]: yeah, renamed to typeParams everywhere. I think it's a bit clearer not to have argTypes and typeArgs 😂

**Pull request merged**